pyomo/contrib/community_detection/detection.py

```diff
@@ -0,0 +1,466 @@
+            community_map[community_key] = (constraint_node_list, variable_node_list)
+
+    elif type_of_community_map == 'c':
+        # If the community map was created for a constraint node graph, then for a given community, we want to create a
+        # new list that contains all of the variables contained in the constraint equations of that community
+
+        for community_key in community_map:
+            constraint_list = sorted(community_map[community_key])
+            variable_list = [constraint_variable_map[numbered_constraint] for numbered_constraint in constraint_list]
+            variable_list = sorted(set([node for variable_sublist in variable_list for node in variable_sublist]))
+            variable_list = [number_component_map[variable] for variable in variable_list]
+            constraint_list = [number_component_map[constraint] for constraint in constraint_list]
+            community_map[community_key] = (constraint_list, variable_list)
+
+    elif type_of_community_map == 'v':
+        # If the community map was created for a variable node graph, then for a given community, we want to create a
+        # new list that contains all of the constraints that the variables of that community appear in
+
+        for community_key in community_map:
+            variable_list = sorted(community_map[community_key])
+            constraint_list = []
+            for numbered_variable in variable_list:
+                constraint_list.extend([constraint_key for constraint_key in constraint_variable_map if
+                                        numbered_variable in constraint_variable_map[constraint_key]])
+            constraint_list = sorted(set(constraint_list))
+            constraint_list = [number_component_map[constraint] for constraint in constraint_list]
+            variable_list = [number_component_map[variable] for variable in variable_list]
+            community_map[community_key] = (constraint_list, variable_list)
+
+    # Thus, each key in community_map now maps to a tuple of two lists, a constraint list and a variable list (in that
+    # order)
+
+    # Log information about the number of communities found from the model
+    logger.info("%s communities were found in the model" % number_of_communities)
+    if number_of_communities == 0:
+        logger.error("in detect_communities: Empty community map was returned")
+    if number_of_communities == 1:
+        logger.warning("Community detection found that with the given parameters, the model could not be decomposed - "
+                       "only one community was found")
+
+    # Return community_map, which has integer keys that now map to a tuple of two lists
+    # containing Pyomo modeling components
+    return community_map
+
+
+def visualize_model_graph(model, community_map=None, type_of_graph='c', with_objective=True, weighted_graph=True,
+                          random_seed=None, type_of_community_map=None, pos=None):
```

Comment:
Why not just have usage:
```
community_map = detect_communities(pyomo_model, ...)
visualize_communities(community_map)
```

Reply:
I'm not sure what you mean - are you saying this function should only take in community_map as an argument?

Reply:
Hi Qi, this actually has to do with the fact that the visualization depends not only on the communities that you derived (which are based on a kind of graph where you applied the decomposition method `type_of_community_map`) but also on the graph you want to plot them into `type_of_graph`. You need the model to derive the original graph shape for plotting.

Reply:
So, what you're telling me is that `detect_communities(pyomo_model,...)` should return a type `CommunityMap` inheriting from `dict` that includes methods for visualization and a `__str__` override?

Reply:
`detect_communities` should return a standard Python `dict`, which is named `community_map`. Visualizing and "stringifying" this `community_map` is done with the functions `visualize_model_graph` and `stringify_community_map` (respectively).

Reply:
@rahuljoglekar47 Apologies, I should be more clear. You're handing the user back a community map from `detect_communities()`. Now, if all the relevant information associated with a community map fits neatly as a python base type (e.g. `dict`), then that's all well and good. However, David is saying that there's additional metadata associated with the communities (i.e. its type). So, under OOP principles, you're no longer just returning a `dict`. You're returning some new object type that has the original `dict` information, but also additional metadata attributes, hence, a new class type.

Reply:
The reason why we thought of `detect_communities(...)` to return a `dict` is that it is the simplest output, independent of the `type_of_graph` selected. We have not considered the general case where we want to return that more complicated object. Let us look into it.